lib/SIL/TypeLowering.cpp

```diff
@@ -2520,3 +2520,68 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
   
   return boxType;
 }
+
+static void countNumberOfInnerFields(unsigned &fieldsCount, SILModule &Module,
+                                     SILType Ty) {
+  if (auto *structDecl = Ty.getStructOrBoundGenericStruct()) {
+    assert(structDecl->hasFixedLayout(Module.getSwiftModule(),
+                                      ResilienceExpansion::Minimal) &&
+           " FSO should not be trying to explode resilient (ie address-only) "
+           "types at all");
+    for (auto prop : structDecl->getStoredProperties()) {
+      SILType propTy = Ty.getFieldType(prop, Module);
+      unsigned fieldsCountBefore = fieldsCount;
+      countNumberOfInnerFields(fieldsCount, Module, propTy);
+      if (fieldsCount == fieldsCountBefore) {
+        // size of Struct(BigStructType) == size of BigStructType()
+        // prevent counting its size as BigStructType()+1
+        ++fieldsCount;
+      }
+    }
+    return;
+  } else if (auto tupleTy = Ty.getAs<TupleType>()) {
+    for (auto elt : tupleTy->getElementTypes()) {
+      auto silElt = SILType::getPrimitiveObjectType(elt->getCanonicalType());
+      countNumberOfInnerFields(fieldsCount, Module, silElt);
+    }
+    return;
+  } else if (auto *enumDecl = Ty.getEnumOrBoundGenericEnum()) {
+    if (enumDecl->isIndirect()) {
+      return;
+    }
+    assert(enumDecl->hasFixedLayout(Module.getSwiftModule(),
+                                    ResilienceExpansion::Minimal) &&
+           " FSO should not be trying to explode resilient (ie address-only) "
+           "types at all");
+    unsigned fieldsCountBefore = fieldsCount;
+    unsigned maxEnumCount = 0;
+    for (auto elt : enumDecl->getAllElements()) {
+      if (!elt->getArgumentInterfaceType()) {
+        continue;
+      }
+      if (elt->isIndirect()) {
+        continue;
+      }
+      // Number of fields > 1 as "future proof" for this heuristic:
+      // In case it is used by a pass that tries to explode enums.
+      // We should probably *never* do that for code size reasons.
+      auto payloadTy = Ty.getEnumElementType(elt, Module);
+      fieldsCount = 0;
+      countNumberOfInnerFields(fieldsCount, Module, payloadTy);
+      if (fieldsCount > maxEnumCount) {
+        maxEnumCount = fieldsCount;
+      }
+    }
+    fieldsCount = fieldsCountBefore + maxEnumCount;
+  }
+}
+
+unsigned TypeConverter::countNumberOfFields(SILType Ty) {
+  unsigned fieldsCount = 0;
+  if (TypeFields.find(Ty) == TypeFields.end()) {
```

Comment:
Can you not recompute here and return early? You are performing two hash table lookups. You do not need to do this. TypeFields.find will return an iterator that will have fieldsCount in it, i.e.:

```
auto Iter = TypeFields.find(Ty);
if (Iter == TypeFields.end()) {
    unsigned fieldsCount = 0;
    countNumberOfInnerFields(fieldsCount, M, Ty);
    TypeFields[Ty] = fieldsCount;
    return std::max(fieldsCount, 1U);
}

return std::max(Iter->second, 1U);
```

Or actually, it would be even cleaner if you flipped the control flow.
```
auto Iter == TypeFields.find(Ty);
if (Iter != TypeFields.end()) {
    return std::max(Iter->second, 1U);
}

unsigned fieldsCount = 0;
countNumberOfInnerFields(fieldsCount, M, Ty);
TypeFields[Ty] = fieldsCount;
return std::max(fieldsCount, 1U);
```